### PR TITLE
feat(api): add `.map` and `.array` accessors

### PR DESCRIPTION
**Pull request by cpcloud**
This PR adds `.map` and `.array` accessors to the `JSONValue` API, which allows quickly accessing map/array methods on weakly typed JSON objects. This was more involved than I initally thought due to snowflake being fairly strict about casting non-object JSON to object.

### CODE REVIEW

File: ibis/expr/types/json.py [hunk 1]
```diff
@@ -17,10 +18,114 @@ class JSONValue(Value):
...
+    @property
+    def map(self) -> ir.MapValue:
+        """Cast this JSON blob to a map of string to JSON.
+
+        Use this property to unlock map functionality on JSON objects.
+
+        Returns
+        -------
+        MapValue
+            Map of string to JSON
```
Comment (cpcloud) [inline ibis/expr/types/json.py:112]: I'd like to add examples here, but these only work with snowflake for the time being.

**Comment by cpcloud** [comment]
Cloud backends look good:

```
…/ibis on  json-attrs is 📦 v4.1.0 via 🐍 v3.10.9 via ❄️  impure (ibis-3.10.9-env) took 5s
❯ pytest -m 'snowflake or bigquery' -n auto -q --dist loadgroup
bringing up nodes...
x.....xx........x........x...x.....x.......xx..x..........xx..x..x....x.....xxx....x.x....xx..x..x.xx...xx...xx..x...xx.xx..x.xx...x..xx........xxxxx.......x..x....x.xx.......x..... [  9%]
....xx.x..xxxxxxxxxx........xx.......x......x.x.....x....xx..x.xx.............x............xxxxxxx.xxx.xxx.xxxxxxxx.x..x.xx.....x.....x..xxxxx....x.x.xx..x.x.....x.x........x....... [ 18%]
..................................x.........x..................................................................x..x...xx...xx.........x..x......x.......xx....x.x.......x............ [ 28%]
..xx..x....x...xxx....x....xx.............x.................x..xxx....x..x.x..........x.x........x..........x..........x.....................................x......................x [ 37%]
................................................................................x.xxxxs..........x............s.xx..x.xs......x.xxx........xxxx.x..x..x....xxx....xx.x.x...x.....x.x. [ 47%]
..x.xx.xxx.x.x..x....x.x......xx....x.x.x..........xx....x....x.xx..x...x.x...xxx.x...x..........x.x..xx.xx.xx...x...x.xx..x.....x..x.xxx.s.......xx........xx...........x.xx..x..x.. [ 56%]
..........x...x........x....x..............x.............x..xx...........x.......x..........xx.....................x........x.....x.............x............x.............x......... [ 66%]
...........x...x..x.x.........xx...............x....x.x..x...xx...x.....x....x....x.........xx.xx.....x...x....xx.......xxx......x......x....x.......x..............x..........x..x.. [ 75%]
.......x............x.............x......x......................................xx.....................x..................................................ss.ss..xxxxx...x.......x... [ 85%]
.......................s.............................................................x....................................................................................xx......... [ 94%]
..s..x.....s...................................x........................x................x.............                                                                               [100%]
1582 passed, 11 skipped, 320 xfailed in 183.39s (0:03:03)
```

File: ibis/expr/types/json.py [hunk 2]
```diff
@@ -17,10 +18,118 @@ class JSONValue(Value):
...
+        ┏━━━━━━━━━━━━━━━━━━━━━┓
+        ┃ JSONGetItem(js, 20) ┃
+        ┡━━━━━━━━━━━━━━━━━━━━━┩
+        │ json                │
+        ├─────────────────────┤
+        │ ∅                   │
+        │ ∅                   │
+        └─────────────────────┘
+        """
         import ibis.expr.operations as ops
```
Comment (kszucs) [inline ibis/expr/types/json.py:99]: You can import ops at the top nowadays :)

Comment (cpcloud) [reply ibis/expr/types/json.py:99]: I think I tried this and it didn't work, let me try again.

File: ibis/backends/snowflake/registry.py [hunk 3]
```diff
@@ -171,9 +194,26 @@ def _unnest(t, op):
     {
         ops.JSONGetItem: fixed_arity(sa.func.get, 2),
         ops.StringFind: _string_find,
-        ops.Map: fixed_arity(sa.func.ibis_udfs.public.object_from_arrays, 2),
-        ops.MapKeys: unary(sa.func.object_keys),
-        ops.MapValues: unary(sa.func.ibis_udfs.public.object_values),
+        ops.Map: fixed_arity(
```
Comment (kszucs) [inline ibis/backends/snowflake/registry.py:197]: Unrelated, but should we consider to use singledispatch for sqlalchemy translation functions?

File: ibis/expr/types/json.py (hunk 2 above)
Comment (cpcloud) [reply ibis/expr/types/json.py:99]: Nice, it works!

**Review by kszucs — APPROVED** [review]
Just nits, otherwise it LGTM.

File: ibis/expr/operations/json.py [hunk 4]
```diff
@@ -14,3 +14,19 @@ class JSONGetItem(Value):
...
+    output_dtype = dt.Array(dt.json)
+    output_shape = rlz.shape_like("args")
+
+
+@public
+class ToJSONMap(Value):
+    arg = rlz.json
+
+    output_dtype = dt.Map(dt.string, dt.json)
+    output_shape = rlz.shape_like("args")
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/expr/operations/json.py]: Done.

File: ibis/backends/snowflake/registry.py (hunk 3 above)
Comment (cpcloud) [reply ibis/backends/snowflake/registry.py:197]: Done.

File: ibis/expr/operations/json.py [hunk 5]
```diff
@@ -14,3 +14,19 @@ class JSONGetItem(Value):
...
     output_dtype = dt.json
     output_shape = rlz.shape_like("args")
+
+
+@public
+class ToJSONArray(Value):
+    arg = rlz.json
+
+    output_dtype = dt.Array(dt.json)
+    output_shape = rlz.shape_like("args")
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/expr/operations/json.py]: Done

File: ibis/backends/snowflake/registry.py (hunk 3 above)
Comment (cpcloud) [reply ibis/backends/snowflake/registry.py:197]: Yeah, we could probably make that work!
